### PR TITLE
VLANs and Interfaces Endpoints

### PR DESCRIPTION
**Pull request by leslitagordita**
* VLANs List
* Linode Create (new interface field)
* Configuration Profile Create (new interface field)
* Configuration Profile Update (new interface field)

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -3291,6 +3291,54 @@ paths:
...
                     items:
                       type: string
+                  interfaces:
+                    type: object
+                    description: >
+                      The Network Interface(s) to add to this Linode. This field is only required when creating
+                      a Linode that is attached to a Virtual Local Area Network (VLAN). **Note**: You must create
+                      the VLAN prior to attaching a Linode to it.
+
+                      * You can add up to 3 Network Interfaces, `eth0` - `eth2`.
```
Comment (aspsk) [inline openapi.yaml:3301]: Please move this below

File: openapi.yaml [hunk 2]
```diff
@@ -3291,6 +3291,54 @@ paths:
...
+
+                      * A Linode can only have one Network Interface of `type: default`. This is the Linode's
+                      public Network Interface, which allows the Linode to communicate across public networks.
+
+                      * The Network Interfaces created by this field are also added to this Linode's
```
Comment (aspsk) [inline openapi.yaml:3306]: Is this phrase not finished?

File: openapi.yaml [hunk 3]
```diff
@@ -3291,6 +3291,54 @@ paths:
...
+                    description: >
+                      The Network Interface(s) to add to this Linode. This field is only required when creating
+                      a Linode that is attached to a Virtual Local Area Network (VLAN). **Note**: You must create
+                      the VLAN prior to attaching a Linode to it.
+
+                      * You can add up to 3 Network Interfaces, `eth0` - `eth2`.
+
+                      * A Linode can only have one Network Interface of `type: default`. This is the Linode's
+                      public Network Interface, which allows the Linode to communicate across public networks.
+
```
Comment (aspsk) [inline openapi.yaml:3305]: * If only `type: additional` interfaces are specified, then this Linode won't have direct access to public Internet. This way you can create truly private Linodes. You can add public interface to the default or a custom boot config later.

File: openapi.yaml [hunk 4]
```diff
@@ -4823,6 +4902,136 @@ paths:
...
+        beta.
+      operationId: addLinodeInterface
+      x-linode-cli-action: interface-create
+      security:
+      - personalAccessToken: []
+      - oauth:
+        - linodes:read_write
+      requestBody:
+        description: >
+          The parameters to set when creating the Configuration profile.
```
Comment (aspsk) [inline openapi.yaml:5001]: Not `Configuration profile`

File: openapi.yaml [hunk 5]
```diff
@@ -4823,6 +4902,136 @@ paths:
...
+          This determines which kernel, devices, how much memory, etc. a Linode boots with.
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Interface'
+      responses:
+        '200':
+          description: |
+            A Configuration profile was created.
```
Comment (aspsk) [inline openapi.yaml:5011]: Interface

File: openapi.yaml [hunk 6]
```diff
@@ -9912,6 +10121,360 @@ paths:
...
+        updates in the future. This notice will be removed when this endpoint is out of
+        beta.
+      operationId: createVlan
+      x-linode-cli-action: vlan-create
+      security:
+      - personalAccessToken: []
+      - oauth:
+        - vlans:read_write
+      requestBody:
+        description: Information about the domain you are registering.
```
Comment (aspsk) [inline openapi.yaml:10211]: domain?

File: openapi.yaml [hunk 7]
```diff
@@ -9912,6 +10121,360 @@ paths:
...
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Vlans'
+      responses:
+        '200':
+          description: NodeBalancer created successfully.
```
Comment (aspsk) [inline openapi.yaml:10219]: Nodebalancer?

File: openapi.yaml [hunk 8]
```diff
@@ -9912,6 +10121,360 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Vlans'
+        default:
+          $ref: '#/components/responses/ErrorResponse'
+      x-code-samples:
+      - lang: Shell
+        source: >
+          curl -H "Authorization: Bearer $TOKEN" \
+                https://api.linode.com/v4beta/networking/vlans/12345
```
Comment (aspsk) [inline openapi.yaml:10291]: would be cool to add output, like
```
{
  "id": 12345,
  "region": "us-east",
  "description": "woo",
  "linodes": [
    {
      "id": 123,
      "mac_address": "ba:ba:b1:54:50:01"
    }
  ],
  "cidr_block": null,
  "created": "2020-06-32T27:47:55"
}
```

File: openapi.yaml [hunk 9]
```diff
@@ -16089,6 +16652,99 @@ components:
...
+          description: This Interface's unique ID.
+          example: 123
+          x-linode-cli-display: 1
+        type:
+          type: string
+          description: |
+            The Network Interface type to create. A Network Interface of type:
+
+            * `default` creates a public Network Interface that accepts external Network traffic. If your Linode
+            does not have a public Network Interface it will not receive a public IP address.
```
Comment (aspsk) [inline openapi.yaml:16702]: (Unless a boot config lists *no* interfaces. In this case for compatibility the Linode will be provided with a public interface.)

File: openapi.yaml [hunk 10]
```diff
@@ -16089,6 +16652,99 @@ components:
...
+          description: |
+           This Network Interface's Media Access Control (MAC) address.
+          example: f2:3c:92:8d:bc:00
+          x-linode-cli-display: 6
+        ip_address:
+          type: string
+          readOnly: true
+          description: |
+           This Network Interface's private IP address. If the Network Interface's type is `Public`, it does not receive an assigned Private IP address.
+          example: f2:3c:92:8d:bc:00
```
Comment (aspsk) [inline openapi.yaml:16746]: 🚫 this is a mac. Example is `10.0.0.123/24`

**Review by aspsk — CHANGES_REQUESTED** [review]
Looks great! Left some minor comments

File: openapi.yaml [hunk 11]
```diff
@@ -17305,6 +17459,68 @@ components:
...
+          type: string
+          minLength: 0
+          maxLength: 64
+          pattern: '/[a-z0-9-]+/'
+          description: |
+            The name of this interface.
+
+            Only used for type `vlan` interfaces. Enter an empty string for type `public` interfaces.
+
+            Interface labels must be unique within the data center region.
```
Comment (alsaihn) [inline openapi.yaml:17477]:
```suggestion
            Interface labels must be unique within this Linode's configuration profile.
```
Customers can and should add the same label to different linode configs, but they must be unique within a config.

File: openapi.yaml [hunk 12]
```diff
@@ -17305,6 +17459,68 @@ components:
...
+          example: example-interface
+        ipam_address:
+          type: string
+          format: ip/netmask
+          description: |
+            This Network Interface's private IP address in Classless Inter-Domain Routing (CIDR) notation.
+
+            Only used for type `vlan` interfaces. Enter an empty string for type `public` interfaces.
```
Comment (alsaihn) [inline openapi.yaml:17485]: Can we note here that an empty string is allowed for the 'vlan' interfaces also? The customer would have to manually configure the network in this case.

File: openapi.yaml [hunk 13]
```diff
@@ -17305,6 +17459,68 @@ components:
...
+
+        **Changes to Linode interface configurations can be enabled by rebooting the Linode with Network Helper enabled.**
+
+        Up to three interfaces can be entered. The position in the array determines the interface to which the settings apply:
+
+        - First/0:  eth0
+        - Second/1: eth1
+        - Third/2:  eth2
+
+        A public interface in the first/eth0 position is automatically assigned the Linode's default public IPv4 address. Only one public interface per Linode can be defined. If no public interface is defined, the Linode will not be reachable via the public internet. Access can only be established via LISH or other Linodes connected by VLAN.
```
Comment (alsaihn) [inline openapi.yaml:17512]: A public interface can be in any position, but I'm not sure about the automatic parts if it's not in position 1.

**Review by alsaihn — COMMENTED** [review]
Could we add something about how adding/changing/removing interfaces is not sufficient by itself to make those changes happen on the linode? A reboot (or other workflow that causes the linode to be booted/restarted) is required for the networking changes to actually take effect. This gives customers the flexibility to make those changes and then have them take effect when convenient for them.

**Review by yoshizzle — APPROVED** [review]
I have the same comments as Carissa. The other things have been addressed.

File: openapi.yaml [hunk 14]
```diff
@@ -17305,6 +17408,73 @@ components:
...
+        ipam_address:
+          type: string
+          format: ip/netmask
+          nullable: true
+          description: |
+            This Network Interface's private IP address in Classless Inter-Domain Routing (CIDR) notation.
+
+            Only used for type `vlan` interfaces. Enter `null` for type `public` interfaces.
+
+            Must be unique among all interfaces in the data center region.
```
Comment (abemassry) [inline openapi.yaml:17439]:
```suggestion
            Must be unique among all interfaces on the Linode.
```

File: openapi.yaml [hunk 15]
```diff
@@ -17305,6 +17408,73 @@ components:
...
+
+            Configuring an IPAM address within the 10.0.0.0/8 block allows for automatic configuration with Network Helper. Addresses outside of this range must be configured manually.
```
Comment (abemassry) [inline openapi.yaml:17441]:
```suggestion
            Configuring an IPAM address within the 10.0.0.0/8 block allows for automatic configuration with Network Helper, for example.
```

File: openapi.yaml [hunk 16]
```diff
@@ -17305,6 +17408,73 @@ components:
...
+
+        **Changes to Linode interface configurations can be enabled by rebooting the Linode with Network Helper enabled.**
+
+        Up to three interfaces can be entered. The position in the array determines the interface to which the settings apply:
+
+        - First/0:  eth0
+        - Second/1: eth1
+        - Third/2:  eth2
+
+        A public interface in the first/eth0 position is automatically assigned the Linode's default public IPv4 address. Only one public interface per Linode can be defined. If no public interface is defined, the Linode will not be reachable via the public internet. Access can only be established via LISH or other Linodes connected by VLAN.
```
Comment (abemassry) [inline openapi.yaml:17466]:
```suggestion
        The the Linode's default public IPv4 address is automatically assigned to the public interface. Only one public interface per Linode can be defined. If no public interface is defined, the Linode will not be reachable via the public internet. Access can only be established via LISH or other Linodes connected by VLAN.
```

File: openapi.yaml [hunk 17]
```diff
@@ -17305,6 +17408,73 @@ components:
...
+
+        Adding a type `vlan` interface attaches this Linode to the labelled VLAN. If the VLAN label is new, a VLAN is created. Up to 10 VLANs can be created in each data center region. You can view your active VLANs using the [/networking/vlans](/docs/api/networking/#vlans-list) endpoint. Only one interface can be assigned to a particular VLAN for each configuration.
+
+        When updating a Linode's interfaces, each interface must be redefined. An empty array results in a default public interface configuration only.
+
+        To detach a Linode from a VLAN, remove the associated interface from the Linode's configuration, delete the configuration, or delete the Linode.
```
Comment (abemassry) [inline openapi.yaml:17472]:
```suggestion
        To detach a Linode from a VLAN, remove the associated interface from the Linode's configuration, delete the configuration, and reboot or delete the Linode.
```

**Review by alsaihn — APPROVED** [review]

**Review by yoshizzle — APPROVED** [review]

**Pull request merged**